maitake-sync/src/blocking/default_mutex.rs

```diff
@@ -0,0 +1,403 @@
+//! Default "chef's choice" [`ScopedRawMutex`] implementation.
+//!
+//! This type is what users will get when they don't override the `Lock` type
+//! parameter for `maitake-sync`'s synchronziation primitives.
+//!
+//! # Notes
+//!
+//! - The `DefaultMutex` cannot ever implement `RawMutex`, only
+//!   `ScopedRawMutex`. This is because it's impossible for a
+//!   `critical-section`-based implementation to implement `RawMutex`, due to
+//!   [`critical-section`'s safety requirements][cs-reqs], which we can't uphold
+//!   in a RAII situation with multiple locks. If we implemented `RawMutex` for
+//!   the non-`critical-section` implementations, then the `critical-section`
+//!   feature flag would *take away* methods that would otherwise be available,
+//!   making it non-additive, which is a BIG NO-NO for feature flags.
+//!
+//! - On the other hand, it *is* okay to have `cfg(loom)` not implement
+//!   `ConstInit` where every other implementation does. This is because `cfg(loom)`
+//!   is a `RUSTFLAGS` cfg rather than a feature flag, and therefore can only be
+//!   enabled by the top-level build. It can't be enabled by a dependency and
+//!   suddenly make your code not compile. Loom users are already used to stuff
+//!   being const-initializable in real life, but not in loom tests, so this is
+//!   more okay.
+//!
+//! [cs-reqs]:
+//!     https://docs.rs/critical-section/latest/critical_section/fn.acquire.html#safety
+#[cfg(not(loom))]
+use super::ConstInit;
+use super::ScopedRawMutex;
+
+/// Default, best-effort [`ScopedRawMutex`] implementation.
+///
+/// This is the default `Lock` type parameter for the [`Mutex`](crate::Mutex)
+/// type, and for the async synchronization primitives that use the blocking
+/// `Mutex`. This type makes a best-effort attempt to Do The Right Thing based
+/// on the currently enabled [feature flags]. In particular, here's what we
+/// currently give you:
+///
+/// - **If `cfg(loom)` is enabled, then the `DefaultMutex` is a [`loom` mutex]**
+///   so that `maitake-sync` primitives work nicely in `loom` tests
+///
+/// - **If the `std` feature is enabled, then the `DefaultMutex` is a
+///   [`std::sync::Mutex`]**, so that `std` users get an OS mutex rather than a
+///   spinlock.
+///
+/// - **If the `critical-section` feature is enabled, then the `DefaultMutex` is
+///   a spinlock that [acquires a critical section][cs] once locked**. This
+///   ensures that bare-metal users who have enabled `critical-section` get a
+///   mutex that disables IRQs when locked.
+///
+/// - **Otherwise, the `DefaultMutex` is a [`Spinlock`]**. This is the default
```

Comment:
I wonder if we could poison this with a `!Sync` bound in this case? That would be enough to inhibit IRQ usage, but would also probably ruin your inter-core use case :/

Reply:
The whole point of a spinlock is to make things `Sync`...if that wasn't the goal, we could just make it panic. I wondered about possibly having some kind of `IrqSafe` marker trait, analogously to `Send` and `Sync`, but...there's no way to actually enforce it...

Reply:
right. whoops.

Reply:
I think the problem is that, although ISRs are "just another kind of concurrency" and can race in all the same ways as threads/multicore code, the things you have to *do* to prevent races are different.